File: app/src/main/java/wseemann/media/jplaylistparser/parser/xspf/XSPFPlaylistParser.java
/*
 * Copyright 2014 William Seemann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package wseemann.media.jplaylistparser.parser.xspf;

import android.util.Log;

import com.yuriy.openradio.gabor.shared.utils.AppLogger;

import org.jdom2.Document;
import org.jdom2.Element;
import org.jdom2.JDOMException;
import org.jdom2.input.SAXBuilder;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Set;

import wseemann.media.jplaylistparser.mime.MediaType;
import wseemann.media.jplaylistparser.parser.AbstractParser;
import wseemann.media.jplaylistparser.playlist.Playlist;
import wseemann.media.jplaylistparser.playlist.PlaylistEntry;

public final class XSPFPlaylistParser extends AbstractParser {

    public final static String EXTENSION = ".xspf";

    private static final String LOCATION_ELEMENT = "LOCATION";
    private static final String TITLE_ELEMENT = "TITLE";
    private static final String TRACK_ELEMENT = "TRACK";
    private static final String TRACKLIST_ELEMENT = "TRACKLIST";

    private static int sNumberOfFiles = 0;

    private static final Set<MediaType> SUPPORTED_TYPES =
            Collections.singleton(MediaType.video("application/xspf+xml"));

    public XSPFPlaylistParser(final int timeout) {
        super(timeout);
    }

    @Override
    public Set<MediaType> getSupportedTypes() {
        return SUPPORTED_TYPES;
    }

    /**
     * Retrieves the files listed in a .asx file
     *
     * @throws IOException
     */
    private void parsePlaylist(final InputStream stream, final Playlist playlist) throws IOException {
        final StringBuilder xml = new StringBuilder();
        String line;
        final BufferedReader reader = new BufferedReader(new InputStreamReader(stream));

        while ((line = reader.readLine()) != null) {
            xml.append(line);
        }

        reader.close();

        parseXML(xml.toString(), playlist);
    }

    private void parseXML(final String xml, final Playlist playlist) {
        final SAXBuilder builder = new SAXBuilder();
        Reader in;
        Document doc;
        Element root;

        try {
            in = new StringReader(xml);

            doc = builder.build(in);
            root = doc.getRootElement();
            final List<Element> children = castList(Element.class, root.getChildren());

            for (int i = 0; i < children.size(); i++) {
                final String tag = children.get(i).getName();

                if (tag != null && tag.equalsIgnoreCase(TRACKLIST_ELEMENT)) {
                    final List<Element> children2 = castList(Element.class, children.get(i).getChildren());

                    for (int j = 0; j < children2.size(); j++) {
                        final String attributeName = children2.get(j).getName();

                        if (attributeName.equalsIgnoreCase(TRACK_ELEMENT)) {
                            final List<Element> children3 = castList(Element.class, children2.get(j).getChildren());
                            buildPlaylistEntry(children3, playlist);
                        }
                    }
                }
            }
        } catch (final JDOMException e) {
            AppLogger.e("XSPF parse exception:" + Log.getStackTraceString(e));
        } catch (final IOException e) {
            AppLogger.e("XSPF parse exception:" + Log.getStackTraceString(e));
        } catch (final Exception e) {
            AppLogger.e("XSPF parse exception:" + Log.getStackTraceString(e));
        }
    }

    private void buildPlaylistEntry(final List<Element> children, final Playlist playlist) {
        final PlaylistEntry playlistEntry = new PlaylistEntry();

        for (int i = 0; i < children.size(); i++) {
            final String attributeName = children.get(i).getName();

            if (attributeName.equalsIgnoreCase(LOCATION_ELEMENT)) {
                String href = children.get(i).getValue();

                // TODO: add trim?
                playlistEntry.set(PlaylistEntry.URI, href);
            } else if (attributeName.equalsIgnoreCase(TITLE_ELEMENT)) {
                String title = children.get(i).getValue();

                if (title != null) {
                    playlistEntry.set(PlaylistEntry.PLAYLIST_METADATA, title);
                }
            }
        }

        sNumberOfFiles = sNumberOfFiles + 1;
        playlistEntry.set(PlaylistEntry.TRACK, String.valueOf(sNumberOfFiles));
        parseEntry(playlistEntry, playlist);
    }

    private <T> List<T> castList(final Class<? extends T> castClass, final List<?> c) {
        final List<T> list = new ArrayList<>(c.size());

        for (final Object o : c) {
            list.add(castClass.cast(o));
        }

        return list;
    }

    @Override
    public void parse(final String uri, final InputStream stream, final Playlist playlist) throws IOException {
        parsePlaylist(stream, playlist);
    }
}

File: app/src/main/java/wseemann/media/jplaylistparser/parser/xspf/XSPFPlaylistParser.kt
/*
 * Copyright 2014 William Seemann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package wseemann.media.jplaylistparser.parser.xspf

import android.util.Log
import com.yuriy.openradio.shared.utils.AppLogger
import org.jdom2.Document
import org.jdom2.Element
import org.jdom2.JDOMException
import org.jdom2.input.SAXBuilder
import wseemann.media.jplaylistparser.mime.MediaType
import wseemann.media.jplaylistparser.parser.AbstractParser
import wseemann.media.jplaylistparser.playlist.Playlist
import wseemann.media.jplaylistparser.playlist.PlaylistEntry
import java.io.IOException
import java.io.InputStream
import java.io.Reader
import java.io.StringReader
import java.util.*

class XSPFPlaylistParser(timeout: Int) : AbstractParser(timeout) {
    /**
     * Retrieves the files listed in a .asx file
     *
     * @throws IOException
     */
    @Throws(IOException::class)
    private fun parsePlaylist(stream: InputStream, playlist: Playlist) {
        val xml = StringBuilder()
        stream.bufferedReader().forEachLine {
            xml.append(it)
        }
        parseXML(xml.toString(), playlist)
    }

    private fun parseXML(xml: String, playlist: Playlist) {
        val builder = SAXBuilder()
        val reader: Reader
        val doc: Document
        val root: Element
        try {
            reader = StringReader(xml)
            doc = builder.build(reader)
            root = doc.rootElement
            val children = castList(Element::class.java, root.children)
            for (i in children.indices) {
                val tag = children[i].name
                if (tag != null && tag.equals(TRACKLIST_ELEMENT, ignoreCase = true)) {
                    val children2 = castList(Element::class.java, children[i].children)
                    for (j in children2.indices) {
                        val attributeName = children2[j].name
                        if (attributeName.equals(TRACK_ELEMENT, ignoreCase = true)) {
                            val children3 = castList(Element::class.java, children2[j].children)
                            buildPlaylistEntry(children3, playlist)
                        }
                    }
                }
            }
        } catch (e: JDOMException) {
            AppLogger.e("XSPF parse exception:" + Log.getStackTraceString(e))
        } catch (e: IOException) {
            AppLogger.e("XSPF parse exception:" + Log.getStackTraceString(e))
        } catch (e: Exception) {
            AppLogger.e("XSPF parse exception:" + Log.getStackTraceString(e))
        }
    }

    private fun buildPlaylistEntry(children: List<Element>, playlist: Playlist) {
        val playlistEntry = PlaylistEntry()
        for (i in children.indices) {
            val attributeName = children[i].name
            if (attributeName.equals(LOCATION_ELEMENT, ignoreCase = true)) {
                val href = children[i].value
                // TODO: add trim?
                playlistEntry[PlaylistEntry.URI] = href
            } else if (attributeName.equals(TITLE_ELEMENT, ignoreCase = true)) {
                val title = children[i].value
                if (title != null) {
                    playlistEntry[PlaylistEntry.PLAYLIST_METADATA] = title
                }
            }
        }
        sNumberOfFiles += 1
        playlistEntry[PlaylistEntry.TRACK] = sNumberOfFiles.toString()
        parseEntry(playlistEntry, playlist)
    }

    private fun <T> castList(castClass: Class<out T>, c: List<*>): List<T> {
        val list: MutableList<T> = ArrayList(c.size)
        for (o in c) {
            list.add(castClass.cast(o))
        }
        return list
    }

    override val supportedTypes: Set<MediaType?>
        get() = setOf(MediaType.video("application/xspf+xml"))

    @Throws(IOException::class)
    override fun parse(uri: String, stream: InputStream, playlist: Playlist) {
        parsePlaylist(stream, playlist)
    }

    companion object {
        const val EXTENSION = ".xspf"
        private const val LOCATION_ELEMENT = "LOCATION"
        private const val TITLE_ELEMENT = "TITLE"
        private const val TRACK_ELEMENT = "TRACK"
        private const val TRACKLIST_ELEMENT = "TRACKLIST"
        private var sNumberOfFiles = 0
    }
}